app/src/main/java/com/asfoundation/wallet/repository/InAppPurchaseService.java

```diff
@@ -19,20 +21,68 @@
 
   private final Repository<String, PaymentTransaction> cache;
   private final ApproveService approveService;
+  private final AllowanceService allowanceService;
   private final BuyService buyService;
   private final BalanceService balanceService;
   private final Scheduler scheduler;
   private final ErrorMapper errorMapper;
 
   public Completable send(String key, PaymentTransaction paymentTransaction) {
-    return checkFunds(key, paymentTransaction, approveService.approve(key, paymentTransaction));
+    return checkFunds(key, paymentTransaction, checkAllowance(key, paymentTransaction));
+  }
+
+  private Completable checkAllowance(String key, PaymentTransaction paymentTransaction) {
+    String fromAddress = paymentTransaction.getTransactionBuilder()
```

Comment:
We could create a local variable of TransactionBuilder to avoid getting it 4 times